File: 2461-maximum-sum-of-distinct-subarrays-with-length-k/2461-maximum-sum-of-distinct-subarrays-with-length-k.py
class Solution:
    def maximumSubarraySum(self, nums: List[int], k: int) -> int:
        
        left = ans = 0
        
        count = defaultdict(int)
        sm = 0
        
        for right in range(len(nums)):
            count[nums[right]] += 1
            sm+=nums[right]
            
            while left < right and count[nums[right]] > 1:
                sm-=nums[left]
                count[nums[left]] -= 1
                left+=1
                   
            if (right - left + 1) == k:
                ans = max(ans, sm)
                sm-=nums[left]
                count[nums[left]]-=1
                left+=1
                

        return ans
                
            
            
            
            
                
        
        
        
        
            
            
            
        
        
            
            
            
            
        
        

File: 2461-maximum-sum-of-distinct-subarrays-with-length-k/2461-maximum-sum-of-distinct-subarrays-with-length-k.cpp
class Solution {
public:
    long long maximumSubarraySum(vector<int>& nums, int k) {
        
        int left =0; long long ans = 0;
        long long sm = 0;
        unordered_map<int, int> count;
        
        for(int right=0; right < nums.size(); ++right){
            
            ++count[nums[right]];
            sm+=nums[right];
            
            while(right < left || count[nums[right]] > 1){
                sm-=nums[left];
                --count[nums[left]];
                ++left;
            }
            
            if(right - left + 1 == k){
                ans = max(sm , ans);
                sm-=nums[left];
                --count[nums[left]];
                ++left;
            }
        }
        
        return ans;
    }
};